### PR TITLE
Add test to track config file changes

### PR DESCRIPTION
**Pull request by tiran**
A new test compares the config from `get_default_config` against a default config file in VCS. This allows us to track any change in the git history.

See: #1725 

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: tests/test_config.py [hunk 1]
```diff
@@ -239,3 +243,31 @@ def test_logging(log_level, debug_level, root, instructlab, openai_httpx):
...
+
+
+
+@patch.multiple(
+    config.DEFAULTS,
+    _cache_home="/cache/instructlab",
+    _config_dir="/config/instructlab",
+    _data_dir="/data/instructlab",
+)
+def test_default_config(testdata_path: pathlib.Path, tmp_path: pathlib.Path, regeneate_testdata: bool):
```
Comment (leseb) [inline tests/test_config.py:255]:
```suggestion
def test_default_config(testdata_path: pathlib.Path, tmp_path: pathlib.Path, regenerate_testdata: bool):
```

File: tests/test_config.py [hunk 2]
```diff
@@ -239,3 +243,31 @@ def test_logging(log_level, debug_level, root, instructlab, openai_httpx):
...
+    current_file = tmp_path / "current_config.yaml"
+
+    current_cfg = config.get_default_config()
+    # roundtrip to verify serialization and de-serialization
+    config.write_config(current_cfg, str(current_file))
+    with current_file.open(encoding="utf-8") as yamlfile:
+        current_content = yaml.safe_load(yamlfile)
+
+    # tox -e py311-unit -- --regenerate-testdata
+    if regeneate_testdata:
```
Comment (leseb) [inline tests/test_config.py:267]:
```suggestion
    if regenerate_testdata:
```

**Review by leseb — COMMENTED** [review]
This looks all good to me. However, I'm wondering if `tests/testdata/default_config.yaml` should be located elsewhere to gain more visibility. If so, how about renaming it to `sample.ilab.config.yaml` or `sample.config.yaml` to indicate that it's a sample configuration file? And have it at the root? or in `src`? Thanks!

File: tests/test_config.py [hunk 3]
```diff
@@ -242,3 +246,35 @@ def test_logging(log_level, debug_level, root, instructlab, openai_httpx):
...
+        current_content = yaml.safe_load(yamlfile)
+
+    if regenerate_testdata:
+        shutil.copy(current_file, saved_file)
+
+    with saved_file.open(encoding="utf-8") as yamlfile:
+        saved_content = yaml.safe_load(yamlfile)
+
+    assert current_content == saved_content, (
+        "current and expected configs are different. If the chance was "
```
Comment (leseb) [inline tests/test_config.py:277]:
```suggestion
        "current and expected configs are different. If the change was "
```

**Comment by leseb** [comment]
> This looks all good to me. However, I'm wondering if `tests/testdata/default_config.yaml` should be located elsewhere to gain more visibility. If so, how about renaming it to `sample.ilab.config.yaml` or `sample.config.yaml` to indicate that it's a sample configuration file? And have it at the root? or in `src`? Thanks!

Discussed offline, we don't want to mix test_data with this proposal, so we will do separately.

**Review by leseb — APPROVED** [review]

**Review by nathan-weinberg — APPROVED** [review]

**Comment by leseb** [comment]
@Mergifyio rebase

**Comment by leseb** [comment]
CI is failing as expected :), I rebased on top of the version flag addition.

```
  E       AssertionError: current and expected configs are different. If the change was intentional, run 'make regenerate-testdata' and commit the updated test data.
  E       assert {'chat': {'context': 'default', 'greedy_mode': False, 'logs_dir': '/data/instructlab/chatlogs', 'max_tokens': None, 'model': '/data/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf', 'session': None, 'vi_mode': False, 'visible_overflow': True}, 'evaluate': {'base_branch': None, 'base_model': 'instructlab/granite-7b-lab', 'branch': None, 'gpus': 1, 'mmlu': {'batch_size': 'auto', 'few_shots': 2}, 'mmlu_branch': {'tasks_dir': '/data/instructlab/datasets'}, 'model': None, 'mt_bench': {'judge_model': 'prometheus-eval/prometheus-8x7b-v2.0', 'max_workers': 16, 'output_dir': '/data/instructlab/internal/eval_data'}, 'mt_bench_branch': {'taxonomy_path': '/data/instructlab/taxonomy'}}, 'general': {'debug_level': 0, 'log_level': 'INFO'}, 'generate': {'chunk_word_count': 1000, 'model': '/data/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf', 'num_cpus': 10, 'output_dir': '/data/instructlab/datasets', 'prompt_file': '/data/instructlab/internal/prompt.txt', 'sdg_scale_factor': 30, 'seed_file': '/data/instructlab/internal/seed_tasks.json', 'taxonomy_base': 'origin/main', 'taxonomy_path': '/data/instructlab/taxonomy'}, 'serve': {'backend': None, 'chat_template': None, 'host_port': '127.0.0.1:8000', 'llama_cpp': {'gpu_layers': -1, 'llm_family': '', 'max_ctx_size': 4096}, 'model_path': '/data/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf', 'vllm': {'llm_family': '', 'vllm_args': []}}, 'train': {'additional_args': {}, 'ckpt_output_dir': '/cache/instructlab/checkpoints', 'data_output_dir': '/data/instructlab/internal', 'data_path': '/data/instructlab/datasets', 'deepspeed_cpu_offload_optimizer': False, 'effective_batch_size': 3840, 'is_padding_free': False, 'lora_quantize_dtype': 'nf4', 'lora_rank': 4, 'max_batch_len': 10000, 'max_seq_len': 4096, 'model_path': 'instructlab/granite-7b-lab', 'nproc_per_node': 1, 'num_epochs': 10, 'save_samples': 250000}, 'version': '1.0.0'} == {'chat': {'context': 'default', 'greedy_mode': False, 'logs_dir': '/data/instructlab/chatlogs', 'max_tokens': None, 'model': '/data/instructlab/models/merlinite-7b-lab-Q4_K_M', 'session': None, 'vi_mode': False, 'visible_overflow': True}, 'evaluate': {'base_branch': None, 'base_model': 'instructlab/granite-7b-lab', 'branch': None, 'gpus': 1, 'mmlu': {'batch_size': 'auto', 'few_shots': 2}, 'mmlu_branch': {'tasks_dir': '/data/instructlab/datasets'}, 'model': None, 'mt_bench': {'judge_model': 'prometheus-eval/prometheus-8x7b-v2.0', 'max_workers': 16, 'output_dir': '/data/instructlab/internal/eval_data'}, 'mt_bench_branch': {'taxonomy_path': '/data/instructlab/taxonomy'}}, 'general': {'debug_level': 0, 'log_level': 'INFO'}, 'generate': {'chunk_word_count': 1000, 'model': '/data/instructlab/models/merlinite-7b-lab-Q4_K_M', 'num_cpus': 10, 'num_instructions': 100, 'output_dir': '/data/instructlab/datasets', 'prompt_file': '/data/instructlab/internal/prompt.txt', 'seed_file': '/data/instructlab/internal/seed_tasks.json', 'taxonomy_base': 'origin/main', 'taxonomy_path': '/data/instructlab/taxonomy'}, 'serve': {'backend': None, 'chat_template': None, 'host_port': '127.0.0.1:8000', 'llama_cpp': {'gpu_layers': -1, 'llm_family': '', 'max_ctx_size': 4096}, 'model_path': '/data/instructlab/models/merlinite-7b-lab-Q4_K_M', 'vllm': {'llm_family': '', 'vllm_args': []}}, 'train': {'additional_args': {}, 'ckpt_output_dir': '/cache/instructlab/checkpoints', 'data_output_dir': '/data/instructlab/internal', 'data_path': '/data/instructlab/datasets', 'deepspeed_cpu_offload_optimizer': False, 'effective_batch_size': 3840, 'is_padding_free': False, 'lora_quantize_dtype': 'nf4', 'lora_rank': 4, 'max_batch_len': 10000, 'max_seq_len': 4096, 'model_path': 'instructlab/granite-7b-lab', 'nproc_per_node': 1, 'num_epochs': 10, 'save_samples': 250000}}
```

**Pull request merged**